src/backend/distributed/planner/recursive_planning.c

```diff
@@ -599,6 +640,337 @@ RecursivelyPlanNonColocatedSubqueriesInWhere(Query *query,
+				 * An OUTER JOIN is recurring if any sides of the join is
+				 * recurring. As in other outer join types, it doesn't matter
+				 * whether the other side was / became recurring or not.
+				 */
+				return leftNodeRecurs || rightNodeRecurs;
+			}
+
+			case JOIN_INNER:
+			{
+				/*
+				 * We don't need to recursively plan non-outer joins and we
+				 * already descended into sub join trees to handle outer joins
+				 * buried in them.
+				 */
+				return leftNodeRecurs && rightNodeRecurs;
+			}
+
+			default:
+			{
+				ereport(ERROR, (errmsg("got unexpected join type (%d) when recursively "
+									   "planning a join",
+									   joinExpr->jointype)));
+			}
+		}
+	}
+	else
+	{
+		ereport(ERROR, errmsg("got unexpected node type (%d) when recursively "
+							  "planning a join",
+							  nodeTag(node)));
+	}
+}
+
+
+/*
+ * RecursivelyPlanDistributedJoinNode is a helper function for
+ * RecursivelyPlanRecurringTupleOuterJoinWalker that recursively plans given
+ * distributed node that is known to be inner side of an outer join.
+ *
+ * We call a node "distributed" if it points to a distributed table or a
+ * more complex object (i.e., a join tree or a subquery) that can be pushed
+ * down to the worker nodes directly. For a join, this means that it's either
+ * an INNER join where any side of it is a distributed table / a distributed
+ * sub join tree, or an OUTER join where the outer side is a distributed table
+ * / a distributed sub join tree.
+ */
+static void
+RecursivelyPlanDistributedJoinNode(Node *node, Query *query,
+								   RecursivePlanningContext *recursivePlanningContext)
+{
+	if (IsA(node, JoinExpr))
+	{
+		/*
+		 * This, for example, means that RecursivelyPlanRecurringTupleOuterJoinWalker
+		 * needs to plan inner side, i.e., "<distributed> INNER JOIN <distributed>",
+		 * of the following join:
+		 *   <recurring> LEFT JOIN (<distributed> INNER JOIN <distributed>)
+		 *
+		 * XXX: Ideally, we should handle such a sub join tree by moving
+		 *      it into a subquery "as a whole" but this implies that we need to
+		 *      rebuild the rtable and re-point all the Vars to the new rtable
+		 *      indexes, so we've not implemented that yet.
+		 *
+		 *      Instead, we recursively plan all the distributed tables in that
+		 *      sub join tree. This is much more inefficient than the other
+		 *      approach (since we lose the opportunity to push-down the whole
+		 *      sub join tree into the workers) but is easier to implement.
+		 */
+
+		RecursivelyPlanDistributedJoinNode(((JoinExpr *) node)->larg,
+										   query, recursivePlanningContext);
+
+		RecursivelyPlanDistributedJoinNode(((JoinExpr *) node)->rarg,
+										   query, recursivePlanningContext);
+
+		return;
+	}
+
+	if (!IsA(node, RangeTblRef))
+	{
+		ereport(ERROR, (errmsg("unexpected join node type (%d)",
+							   nodeTag(node))));
+	}
+
+	RangeTblRef *rangeTableRef = (RangeTblRef *) node;
+	if (IsRTERefRecurring(rangeTableRef, query))
+	{
```

Comment:
minor: 1 line comment would be good before return:
/* there is no need to recursively plan already recurring RTEs */